iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+    auto operator=(const AttachmentId& rhs) -> AttachmentId& = delete;
+
+    AttachmentId(AttachmentId&& rhs) noexcept;
+    auto operator=(AttachmentId&& rhs) noexcept -> AttachmentId&;
+    ~AttachmentId();
+
+    /// Creates an [`AttachmentId`] from a [`Guard`] that was returned via
+    /// [`WaitSet::attach_interval()`], [`WaitSet::attach_notification()`] or
+    /// [`WaitSet::attach_deadline()`].
+    static auto from_guard(const Guard<S>& guard) -> AttachmentId;
+
+    /// Returns true if an event was emitted from a notification or deadline attachment
+    /// corresponding to [`Guard`].
+    auto event_from(const Guard<S>& guard) const -> bool;
+
+    /// Returns true if the deadline for the attachment corresponding to [`Guard`] was missed.
+    auto deadline_from(const Guard<S>& guard) const -> bool;
+
+  private:
+    explicit AttachmentId(iox2_attachment_id_h handle);
+    template <ServiceType>
+    friend auto run_callback(iox2_attachment_id_h, void*);
+    template <ServiceType ST>
+    friend auto operator==(const AttachmentId<ST>&, const AttachmentId<ST>&) -> bool;
+    template <ServiceType ST>
+    friend auto operator<(const AttachmentId<ST>&, const AttachmentId<ST>&) -> bool;
+
+    void drop();
+
+    iox2_attachment_id_h m_handle = nullptr;
+};
+
+template <ServiceType S>
+auto operator==(const AttachmentId<S>& lhs, const AttachmentId<S>& rhs) -> bool;
+
+template <ServiceType S>
+auto operator<(const AttachmentId<S>& lhs, const AttachmentId<S>& rhs) -> bool;
+
+/// The [`WaitSet`] implements a reactor pattern and allows to wait on multiple events in one
+/// single call [`WaitSet::run_once()`] until it wakes up or to run repeatedly with
+/// [`WaitSet::run()`] until the a interrupt or termination signal was received or the user
+/// has explicitly requested to stop with [`WaitSet::stop()`].
+///
+/// The [`Listener`] can be attached as well as sockets or anything else that
+/// can be packed into a [`FileDescriptorView`].
+///
+/// Can be created via the [`WaitSetBuilder`].
+template <ServiceType S>
+class WaitSet {
+  public:
+    WaitSet(const WaitSet&) = delete;
+    auto operator=(const WaitSet&) -> WaitSet& = delete;
+    WaitSet(WaitSet&&) noexcept;
+    auto operator=(WaitSet&&) noexcept -> WaitSet&;
+    ~WaitSet();
+
+    /// Can be called from within a callback during [`WaitSet::run()`] to signal the [`WaitSet`]
+    /// to stop running after this iteration.
+    void stop();
+
+    /// Waits in an infinite loop on the [`WaitSet`]. The provided callback is called for every
+    /// attachment that was triggered and the [`AttachmentId`] is provided as an input argument to
+    /// acquire the source.
+    /// If an interrupt- (`SIGINT`) or a termination-signal (`SIGTERM`) was received, it will exit
+    /// the loop and inform the user via [`WaitSetRunResult`].
+    auto run(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<WaitSetRunResult, WaitSetRunError>;
```

Comment:
I really don't like the name `run` (as mentioned in the prior PR). What is "running"? The comment even explains that it is "waiting in an infinite loop". @elBoberido

Reply:
@elBoberido I would side here with the native speaker @orecham if you do not have any counter arguments.

So I would rename:
 * `run` -> `wait`
 * `run_once` -> `try_wait`

Reply:
@elBoberido Doesn't the return of a `WaitSetRunResult` tell this story sufficiently?

Reply:
I don't think so. Playing the devils advocate, it should be `WaitSetWaitResult` to be consistent with the function name :).
But even if the name would remain the same, it think it would not be sufficient on the call site, e.g.
```rs
waitset.wait(|id| { ... })?;
waitset.wait(on_event)?;
waitset.wait(event_handler)?;
```
In this case, only `on_event` would give a hint. But even then, it would be ambiguous since one could assume to wait until the condition in `on_event` becomes true and then return, like with an condition variable. For myself, I wouldn't call a free function `on_event` since that describes how it is used, not what it does. For a parameter name it would be fine, though.
There is also the familiarity with other APIs with just a `wait` fuction. These are usually just blocking and do not perform actions.
I've the feeling, there is still the iceoryx1 WaitSet in our minds, which indeed just blocks and returns a list of events/states. The current WaitSet exceeds this functionality and if the attach methods would have a callback for each event itself, it would be known as an EventDispatcher in the literature. Currently it's something in between an EventDispatcher and what we know of in iceoryx1. I guess this is the reason why it's so difficult to find a proper name.
Also, keep in mind that the cose is read much more often than written, therefore one should optimize for the reader, not the writer of the code

Reply:
I can settle for the `wait_and_process_events` middleground.

Reply:
Some suggestions from chatgpt, in case we want a shorter name
- wait_and_process
- process_events
- await_and_process

Reply:
Argh, I read it too late.